File: js/euler49.js
function euler49(){
    var primes = new Set();
    var sieved = new Array(10000).fill(false);
    for (let i = 2; i <= sieved.length; i++){
        if (!sieved[i]){
            for (let j = 2*i; j < sieved.length; j += i){
                sieved[j] = true;
            }
            if (i >= 1000) primes.add(i);
        }
    }
    var map = new Map();
    for (let prime of primes){
        let digits = prime.toString().split('').sort().join('');
        if (digits == "1478") continue;
        if (!map.has(digits)){
            map.set(digits, new Set());
        }
        map.get(digits).add(prime);
    }
    var res = "";
    for (let val of map.values()){
        if (val.size < 3) continue;
        res = getSequence(val);
        if (res != "") break;
    }
    return res;
}

function getSequence(nums){
    let seq = Array.from(nums);
    for (let i = 0; i + 2 < seq.length; i++){
        for (let j = i + 1; j < seq.length; j++){
            let diff = seq[j] - seq[i];
            if (nums.has(seq[i] + 2 * diff)){
                return "" + seq[i] + seq[j] + (seq[i] + 2 * diff);
            }
        }
    }
    return "";
}

console.log(euler49());

File: go/euler49.go
package main
import (
    "fmt"
    "sort"
    "strings"
    "strconv"
)

func buildGroups() map[string][]int {
    groups := make(map[string][]int)
    for n := 1001; n <= 9999; n += 2 {
        if isPrime(n) {
            digits := sortDigits(n)
            
            if (digits == "1478") {
                continue
            }
            _, found := groups[digits]
            if !found {
                groups[digits] = []int{}
            }
            groups[digits] = append(groups[digits], n)
        }
    }
    return groups
}

func isPrime(n int) bool {
    if n < 2 {
        return false
    }
    for i := 2; i * i <= n; i++ {
        if n % i == 0 {
            return false
        }
    }
    return true
}

func buildSequence(nums []int) int {
    seen := make(map[int]bool)
    for _, num := range nums {
        seen[num] = true
    }
    for j := 1; j < len(nums); j++ {
        for i := 0; i < j; i++ {
            diff := nums[j] - nums[i]
            third := nums[j] + diff
            _, found := seen[third]
            if found {
                return nums[i] * 100000000 + nums[j] * 10000 + third
            }
        }
    }
    
    return 0
}

func sortDigits(n int) string {
    s := strings.Split(strconv.Itoa(n), "")
    sort.Strings(s)
    digits := strings.Join(s, "")
    return digits
}

func main() {
    res := 0
    groups := buildGroups()
    for _, group := range groups{
        if len(group) >= 3 {
            res = buildSequence(group)
            if res != 0 {
                break
            }
        }
    }
    fmt.Println(res)
}
